### DISCUSSION TITLE
onUploadProgress

### DISCUSSION THREAD

**Discussion by alsherif-khalaf** [directus/directus#13882] [Archive]
Hi ,

transport [optional] Object - Defines settings you want to customize regarding [transport](https://docs.directus.io/reference/sdk/#transport).
params [optional] Object - Defines an object with keys and values to be passed as additional query string.
headers [optional] Object - Defines an object with keys and values to be passed as additional headers.
onUploadProgress [optional] (event: [ProgressEvent (opens new window)](https://developer.mozilla.org/en-US/docs/Web/API/ProgressEvent)=> void) - Defines a callback function to indicate the upload progress.

https://docs.directus.io/reference/sdk/

How can I use the onUploadProgress 

is there any example ?

``
 const directus = new Directus(url, onUploadProgress)
const response = await directus.files.createMany(imagesFormData).onUploadProgress(setProgress(Math.round((100 * data.loaded) / data.total)))

``

ReferenceError: onUploadProgress is not defined

**Comment by br41nslug** [comment]
To use the onUploadProgress function you'll need to hook into it like so:
```js
import { Directus } from "@directus/sdk";

export const directus = new Directus(process.env.REACT_APP_DIRECTUS_URL, {
    transport: {
        onUploadProgress: (evt) => {
            console.log('progress', evt)
        },
    }
});

const form = document.querySelector('#upload-file');

if (form) {
    form.addEventListener('submit', async (event) => {
        event.preventDefault();

        const form = new FormData(event.target);
        await directus.files.createOne(form);
    });
}
```

**Reply by alsherif-khalaf** [reply]
It worked like a charm 
thank you :)

**Reply by zguig52** [reply]
Hi there,

I saw that the SDK changed and couldn't find a way to easily track upload progress status like was advised in this discussion. Basically I need this to inform users during files upload.

Would somebody have a simple code snippet to perform this with the SDK?

From what I understood, there might be something to do with polyfilling and overwriting the fetch API via globals.
https://docs.directus.io/guides/sdk/getting-started.html
https://github.com/directus/directus/discussions/19458

But I couldn't find any fetch lib progress available yet:
https://github.com/unjs/ofetch/issues/45

Is there something simple to allow this or should I manually use Axios or write an XMLHttpRequest as is discussed here?
https://stackoverflow.com/questions/35711724/upload-progress-indicators-for-fetch

**Reply by br41nslug** [reply]
The `fetch` API does not support uploadprogress handlers like the old XMLHttpRequest API.
Had some hope you could use `ofetch` for that but as you've found the issue to implement that is still open and looks stale.

So currently the only way you could make this work is to write a wrapper for `axios` (or plain XMLHttpRequest) making it compatible with the `fetch` API so you can use it as `fetch` replacement in the SDK. But realistically if this is a "must have" feature you're probably better off using the old SDK in terms of engineering lift.

**Reply by zguig52** [reply]
@br41nslug, thanks for your reply, I will see what I can do to be able uploading large files and having the upload status and will come back here to provide a workaround with new SDK if I manage to do something :).

**Comment by zguig52** [comment]
Hereafter a small code snippet to allow uploading with XMLHttpRequest when progressStatus is required and use SDK if not, while preserving the same function signature:

```
export const uploadCMSFiles = async (
  files: File[],
  folder?: string,
  onUploadProgress?: Function,
) => {
  if(files.length > 0){
    const formData = new FormData();
    for (let file of files) {
      if(folder && folderFound){
        formData.append("folder", folder);
      }
      formData.append("file", file);
    }
    try{
      let response: any;
      if(onUploadProgress){
        const xhr = new XMLHttpRequest();
//Here to be put Directus API root URL
        const url = useRuntimeConfig().public.apiURL as string;
//Here to be put Directus current token to be used if authentication required
        const authHeader = useAuth().token.value;
        response = await new Promise((resolve) => {
          xhr.upload.addEventListener("progress", (event) => onUploadProgress(event));
          xhr.addEventListener("progress", (event) => onUploadProgress(event));
          xhr.open("POST", url + "/files", true);
          xhr.onload = function(event) {
            resolve({data: JSON.parse(xhr.responseText)?.data, status: xhr.status});
          };
          if(authHeader){
            xhr.setRequestHeader("Authorization", authHeader);
          }
          xhr.send(formData);
        });
        if(response && response.status == 200){
          response = response.data;
        }else{
          throw("Cannot upload file: " + response.status + " / " + response);
        }
      }else{
//Classic directus sdk client usage
        let client = directusAuthenticatedRestClient();
        response = await client.request(sdkUploadFiles(formData));
      }
      if(files.length == 1){
        return [response] as DirectusFile[];
      }else{
        return response as DirectusFile[];
      }
    }catch(error: any){
      console.error("uploadCMSFiles", error);
      displayMessage("error", error.errors ? error.errors[0]?.message : error);
    }
  }
}
```
